<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" >

    <xacro:arg name="sim_mode" default="false" />
   
    <!-- GAZEBO CONTROL MANAGER (per simulazione) -->
    <xacro:if value="$(arg sim_mode)">
        <ros2_control name="GazeboSystem" type="system">
            <hardware>
                <plugin>gazebo_ros2_control/GazeboSystem</plugin>
            </hardware>
            <joint name="base_left_wheel_joint">
                <command_interface name="velocity">
                    <param name="min">-10</param>
                    <param name="max">10</param>
                </command_interface>
                <state_interface name="position" />
                <state_interface name="velocity" />
            </joint>
            <joint name="base_right_wheel_joint">
                <command_interface name="velocity">
                    <param name="min">-10</param>
                    <param name="max">10</param>
                </command_interface>
                <state_interface name="velocity" />
                <state_interface name="position" />
            </joint>
        </ros2_control>
    </xacro:if>

    <!-- Controllers per Gazebo -->
    <xacro:if value="$(arg sim_mode)">
        <gazebo>
            <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
                <parameters>$(find robot_ros)/config/my_controllers.yaml</parameters>
                <parameters>$(find robot_ros)/config/gaz_ros2_ctl_sim.yaml</parameters>
            </plugin>
        </gazebo>
    </xacro:if>

    <!-- REAL CONTROL MANAGER (per robot reale) -->
    <xacro:unless value="$(arg sim_mode)">
        <ros2_control name="RealRobot" type="system">
            <hardware>
                <plugin>arduino_hardware_interface/ArduinoHardwareInterface</plugin>
                <param name="device">/dev/ttyACM0</param>
                <param name="baud_rate">57600</param>
                <param name="timeout">1000</param>
                <param name="enc_counts_per_rev">3436</param>
                <param name="wheel_radius">0.065</param>
                <param name="wheel_separation">0.17</param>
            </hardware>
            
            <joint name="base_left_wheel_joint">
                <command_interface name="velocity">
                    <param name="min">-10</param>
                    <param name="max">10</param>
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/>
            </joint>
            <joint name="base_right_wheel_joint">
                <command_interface name="velocity">
                    <param name="min">-10</param>
                    <param name="max">10</param>
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/>
            </joint>
        </ros2_control>
    </xacro:unless>
  
</robot>

